include/tscore/AtomicBit.h

```diff
@@ -0,0 +1,92 @@
+/**
+  @file AtomicBit
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  @section details Details
+
+////////////////////////////////////////////
+  Implements class AtomicBit
+*/
+
+#pragma once
+#include "stdint.h"
+#include <atomic>
+
+//////////////////////////////////////////////////////
+/// AtomicBit for inplace atomic bit operations
+/* useful when you refernce a packed `bit` as a `bool&`,
+ * you want a bit to 'walk and talk' like an std::atomic<bool> or std::atomic_flag.
+ * In practice this is constructed at time of the operation(s),
+ * storing it would defeat the purpose of packing the bits.
+ */
+class AtomicBit
+{
+  std::atomic<uint8_t> *_byte_ptr; ///< pointer to the byte
+  uint8_t const _mask;             ///< bitmask of which bit you are using
+
+public:
+  // define a bit to perform atomic operations
+  AtomicBit(uint8_t *byte_ptr, const uint8_t mask) : _byte_ptr(reinterpret_cast<std::atomic<uint8_t> *>(byte_ptr)), _mask(mask) {}
+
+  // Atomically set the bit true
+  // return true if the bit was false
+  // return false if the bit was true
+  bool
+  test_and_set()
+  {
+    return compare_exchange(true);
+  }
+
+  // allows assign by bool
```

Comment:
"@return The previous value of the bit."

Reply:
actually, this "@return The new value of the bit." so that expressions like, 
`a = b = c = true;` would set all values to true regardless if previous state.  Not that I reccomend that assignment style.